include/sycl_ext_complex.hpp

```diff
@@ -1518,10 +1518,559 @@ operator<<(const sycl::stream &__ss, const complex<_Tp> &_x) {
+      rtn[i] = lhs op rhs[i];                                                  \
+                                                                               \
+    return rtn;                                                                \
+  }
+
+  OP(==)
+  OP(!=)
+
+#undef OP
+
+  // OP is: <, >, <=, >=
+#define OP(op)                                                                 \
+  friend marray<bool, NumElements> operator op(const marray &lhs,              \
+                                               const marray &rhs) = delete;    \
+  friend marray<bool, NumElements> operator op(const marray &lhs,              \
+                                               const DataT &rhs) = delete;     \
+  friend marray<bool, NumElements> operator op(const DataT &lhs,               \
+                                               const marray &rhs) = delete;
+
+  OP(<);
+  OP(>);
+  OP(<=);
+  OP(>=);
+
+#undef OP
+
+  friend marray operator~(const marray &v) = delete;
+
+  friend marray<bool, NumElements> operator!(const marray &v) = delete;
+};
+
+_SYCL_END_NAMESPACE
+
+_SYCL_EXT_CPLX_BEGIN_NAMESPACE
+
+// Make_complex_marray
+
+template <class T, std::size_t NumElements>
+sycl::marray<complex<T>, NumElements>
+make_complex_marray(const sycl::marray<T, NumElements> &real,
+                    const sycl::marray<T, NumElements> &imag) {
+  sycl::marray<complex<T>, NumElements> rtn;
+
+  for (std::size_t i = 0; i < NumElements; ++i) {
+    rtn[i].real(real[i]);
+    rtn[i].imag(imag[i]);
+  }
+
+  return rtn;
+}
+
+template <class T, std::size_t NumElements>
+sycl::marray<complex<T>, NumElements>
+make_complex_marray(const sycl::marray<T, NumElements> &real, const T &imag) {
+  sycl::marray<complex<T>, NumElements> rtn;
+
+  for (std::size_t i = 0; i < NumElements; ++i) {
+    rtn[i].real(real[i]);
+    rtn[i].imag(imag);
+  }
+
+  return rtn;
+}
+
+template <class T, std::size_t NumElements>
+sycl::marray<complex<T>, NumElements>
+make_complex_marray(const T &real, const sycl::marray<T, NumElements> &imag) {
+  sycl::marray<complex<T>, NumElements> rtn;
+
+  for (std::size_t i = 0; i < NumElements; ++i) {
+    rtn[i].real(real);
+    rtn[i].imag(imag[i]);
+  }
+
+  return rtn;
+}
+
+template <class T, std::size_t NumElements, std::size_t... I>
+auto constexpr make_complex_marray(
+    const sycl::marray<complex<T>, NumElements> &cmplx,
+    std::integer_sequence<std::size_t, I...> int_seq) {
+  return sycl::marray<complex<T>, int_seq.size()>{cmplx[I]...};
+}
+
+template <class T, std::size_t NumElements, std::size_t... I>
+auto constexpr make_complex_marray(
+    const sycl::marray<T, NumElements> &real,
+    const sycl::marray<T, NumElements> &imag,
+    std::integer_sequence<std::size_t, I...> int_seq) {
+  return sycl::marray<complex<T>, int_seq.size()>{
+      complex<T>(real[I], imag[I])...};
+}
+
+// Get
+
+template <class T, std::size_t NumElements>
```

Comment:
Should we have those free functions?
Why nover overlead `.real()` on marray to return those?

Reply:
I can also remove the Getter and Setter.
There is no ```real``` neither ```imag``` methods for the specialisation of the marray<complex>

Would you prefer not to have these functionalities on marray (just delete it)
Or implement ```real``` and ```imag``` methods on the marray specialization?

```
real() -> sycl::marray<T, NumElements> // getter
imag() -> sycl::marray<T, NumElements> // getter

real(sycl::marray<T, NumElements>) -> sycl::marray<T, NumElements> // setter
imag(sycl::marray<T, NumElements>) -> sycl::marray<T, NumElements> // setter
```

Reply:
@TApplencourt ?

Reply:
> Or implement real and imag methods on the marray specialization?
This sound cleaner to me!

(sorry for the delay)

Reply:
Here's the list of all getters and setters proposed in this PR:

```
// getter functions
sycl::marray<T, NumElements> real();
sycl::marray<T, NumElements> imag();
T real(std::size_t index);
T imag(std::size_t index);
template <std::size_t... I> auto constexpr real(std::integer_sequence<std::size_t, I...> int_seq);
template <std::size_t... I> auto constexpr imag(std::integer_sequence<std::size_t, I...> int_seq)

// setter functions
void real(const sycl::marray<T, NumElements> &values);
void imag(const sycl::marray<T, NumElements> &values);
void real(const T &value);
void imag(const T &value);
void real(std::size_t index, const T &value);
void imag(std::size_t index, const T &value);
```

One problem with overloading the ```real``` and ```imag``` for both getters and
setters is that we will create an ambiguous situation since overload resolution
doesn't depend on return values, but only arguments.

- getter: ```T real(std::size_t index); ```
- setter: ```void real(const T &value);```

What we could do is to keep the ```getX``` and ```setX``` and overload them, or
other variants like implicit (ex: ```real```) for getter and explicit
(ex: ```setReal```) for setters.

What do you think?

Reply:
I think keeping `real() / imag()` as a getter is a good idea (it mimic `std::complex` )
By the way, do we need  `real(std::size_t index)` I guess we can always do `a[i].real()`?


For setting, this is a good question... Can we just do like std::complex?
Aka just use `=` ?  

If the user passes a `std::marray<complex>` we set both real and imaging, and if just a `std::marrau<float>` we set just the real part? If people want to set a particular value `a[i]=23` should do the trick. 

What do you think?

Reply:
Since we are starting to think about the design of the marray implementation and how to change it, I believe we should do this as part of the DPC++ and Khronos proposal.

I'll update the marray's DPC++ proposal, and I'll ping you there so we can continue our discussion together with the community as well.

Then, we'll take back the decision made in the proposal and I'll update the implementation

Reply:
Good idea; thanks a lot!